### PR TITLE
[1/2] Add watershed documentation

### PR DESCRIPTION
**Pull request by ambrosejcarr**
This PR adds documentation but does not significantly modify the watershed code. 

@dganguli can you take a look at the comments and verify that what I've written represents what this code should be doing?

### CODE REVIEW

**Comment by ambrosejcarr** [comment]
Per offline conversation with @dganguli , the stain parameter (segmentation mask) determines the extent to which segmentation obtains cells vs nuclei. 

The parameters in the ISS notebook are derived from culture segmentation, and don't work well here. I should adjust the text to reflect these facts, and that this is a cell segmentation algorithm, not a nuclei one!

File: starfish/image/_segmentation/watershed.py [hunk 1]
```diff
@@ -12,15 +12,37 @@
...
-    def __init__(self, dapi_threshold, input_threshold, min_distance, **kwargs) -> None:
+
+    def __init__(
+        self,
+        dapi_threshold: Number,
+        input_threshold: Number,
+        min_distance: int,
+        **kwargs
+    ) -> None:
+        """Implements watershed segmentation of cells seeded from a DAPI image the the point clouds
```
Comment (dganguli) [inline starfish/image/_segmentation/watershed.py]: Implements watershed segmentation of cells. Algorithm is seeded by DAPI image. Binary segmentation mask is computed from a maximum projection of spots across C and R, which is subsequently thresholded.

File: starfish/image/_segmentation/watershed.py [hunk 2]
```diff
@@ -64,76 +103,217 @@ def show(self, figsize: Tuple[int, int]=(10, 10)) -> None:
...
         self.markers, self.num_cells = self.label_nuclei(
+            self.dapi_thresholded,
             min_allowed_size, max_allowed_size, min_dist
         )
         self.mask = self.watershed_mask(stain_thresh, self.markers, disk_size_mask)
         self.segmented = self.watershed(self.markers, self.mask)
         return self.segmented
 
-    def filter_dapi(self, dapi_thresh, disk_size):
+    def filter_dapi(self, dapi_thresh: float, disk_size: Optional[int]) -> np.ndarray:
```
Comment (dganguli) [inline starfish/image/_segmentation/watershed.py]: should be called filter_nuclei. in general, we should not refer to dapi, as there are other ways to stain nuclei, e.g., drac5 as employed by Dartfish.

File: starfish/image/_segmentation/watershed.py [hunk 3]
```diff
@@ -12,15 +12,37 @@
...
+        min_distance: int,
+        **kwargs
+    ) -> None:
+        """Implements watershed segmentation of cells seeded from a DAPI image the the point clouds
+
+        Watershed segmentation proceeds by constructing basins that extend from the DAPI image
+        and the point cloud constructed from a maximum projection across rounds and channels. It
+        constructs a mask that extends from the basins to prevent non-cellular regions from being
+        included.
+
```
Comment (dganguli) [inline starfish/image/_segmentation/watershed.py]: add: this simply wraps skimage. blah blah blah dot watershed and link to that functions documentation as well?

**Pull request merged**